cdap-ui/app/cdap/components/NamespaceDetails/Preferences/index.js

```diff
@@ -41,130 +46,205 @@ const PREFERENCES_TABLE_HEADERS = [
   }
 ];
 
-const renderPreferencesTableBody = (prefs) => {
-  return (
-    <tbody>
-      {
-        prefs
-          .map((pref, index) => {
-            return (
-              <tr key={index}>
-                <td title={pref.key}>
-                  {pref.key}
-                </td>
-                <td>{pref.scope}</td>
-                <td title={pref.value}>
-                  {pref.value}
-                </td>
-              </tr>
-            );
-          })
+class NamespaceDetailsPreferences extends Component {
+  state = {
+    modalOpen: false,
+    namespacePrefs: this.props.namespacePrefs,
+    systemPrefs: this.props.systemPrefs,
+    newNamespacePrefsKeys: [],
+    viewAll: false
+  };
+
+  static propTypes = {
+    namespacePrefs: PropTypes.object,
+    systemPrefs: PropTypes.object,
+  };
+
+  eventEmitter = ee(ee);
+
+  componentDidMount() {
+    this.eventEmitter.on(globalEvents.NSPREFERENCESSAVED, getNamespacePrefs);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    let newNamespacePrefsKeys = [];
+    let namespacePrefs = this.state.namespacePrefs;
+
+    Object.keys(nextProps.namespacePrefs).forEach(nsPrefKey => {
+      if (!(nsPrefKey in namespacePrefs) || (nsPrefKey in namespacePrefs && namespacePrefs[nsPrefKey] !== nextProps.namespacePrefs[nsPrefKey])) {
+        newNamespacePrefsKeys.push(nsPrefKey);
       }
-    </tbody>
-  );
-};
+    });
 
-const convertPrefsArray = (prefs, scope) => {
-  return Object.keys(prefs).map(prefKey => {
-    return {
-      key: prefKey,
-      scope,
-      value: prefs[prefKey]
-    };
-  });
-};
+    let viewAll = this.state.viewAll;
 
-const mapStateToProps = (state) => {
-  return {
-    namespacePrefs: state.namespacePrefs,
-    systemPrefs: state.systemPrefs
+    // If a newly added/edited pref is not within the first 5 prefs alphabetically,
+    // then we have to show all prefs to be able to highlight the new pref
+    let newPrefs = {...nextProps.namespacePrefs, ...this.state.systemPrefs};
```

Comment:
if the key to the preference in namespace and system is the same, we still want to show both? This will override the namespace preference with the system preference

Reply:
Hmm not sure if we want to show both or not, will check with @leacd

Reply:
Confirmed with Lea, we want to override the system preference with the namespace one, so I've flipped the object spread here.